### ISSUE TITLE
Writing on HttpServerResponse does not throw when closed

### ISSUE THREAD

**Issue by FroMage** [eclipse/vert.x#2181]
When the client closes the connection, I can keep on calling HttpServerResponse.write forever without any indication of an error, I don't even get an exception in my exception handler.

It would be better to throw something.

**Comment by vietj** [comment]
I agree it should be done.

**Comment by glassfox** [comment]
Sharp solution - now I have millions of Error logs. I propose to make some flag that enable this feature or disable it. This will allow silently complete the action without stress the application.

**Comment by torahmike** [comment]
Strongly disagree with this change. I don't want a client disconnecting to be a SEVERE exception. This happens all the time when a user quickly opens and closes a web page.

**Comment by vietj** [comment]
@torahmike the main issue is that your application might think it is sending a correct response to the user and actually it is not

**Comment by FroMage** [comment]
Why can't you catch the exception and decide for yourself if it's important or not?

**Comment by torahmike** [comment]
Because it's caught deep inside the library before it reaches application
code.

**Comment by torahmike** [comment]
> @torahmike the main issue is that your application might think it is sending a response to the user and actually it is not.

@vietj Typical Java guidelines are to only use Exceptions for exceptional situations, not routine ones like a client disconnecting. Java exceptions are expensive.

**Comment by vietj** [comment]
@torahmike you are wrong, this exception is not sent because a client has disconnected, this exception is sent because your application is writing a response to a client and it might think that the client will receive the response and it will not, so there shall be a way to make the application aware of this

**Comment by tsegismont** [comment]
@vietj your are right, there shall be a way to make the application aware of this.

The question is: was this change the right way to do it? Considering the feedback we had so far from production users, I believe it wasn't.

**Comment by vietj** [comment]
we had some feedback, it was not huge either.

however I'm willing to address this but first find what should be the correct thing for Vert.x 4 and the retrofit that to 3.x

**Comment by torahmike** [comment]
> @torahmike you are wrong, this exception is not sent because a client has disconnected, this exception is raised because your application is writing a response to a client and it might think that the client will receive the response and it will not, so there shall be a way to make the application aware of this

@vietj It's thrown whenever a client disconnects before a response is sent, which is a very routine occurrence, in my experience.

**Comment by vietj** [comment]
yes it is.

however I think we are dealing with a trade off between usability and reliability.

the big question is what do we want to do.

What I would like to do is:

- remove the exception in write(Buffer buffer)
- add a write(Buffer buffer, Handler<AsyncResult<Void> handler) that would get this exception

however this does not play well with Vert.x 4 future based design because it means that the write method should return Future<Void>.

And this is where we have an issue because the HttpServerResponse#write(Buffer) method is fluent and is already returning HttpServerResponse.

So to summarise I want to improve the usability but there is a trade off (sacrifice) to do.
